File: apps/desktop/src/renderer/src/utils/adaptivePerformance.ts
import type { DesktopPerformanceResourcesSnapshot } from '../../../shared/protocol';
import type { DiagnosticsSnapshot } from '../services/documentSession';

export type AdaptivePerformanceLevel = 0 | 1 | 2 | 3;

export interface AdaptivePerformanceSnapshot {
  readonly level: AdaptivePerformanceLevel;
  readonly detectedRefreshHz: number;
  readonly targetFrameMs: number;
  readonly p95FrameMs: number;
  readonly p95InputLatencyMs: number;
  readonly framePressure: number;
  readonly renderPressure: number;
  readonly resourcePressure: number;
  readonly reason: 'headroom' | 'frame-time' | 'input-latency' | 'render-backlog' | 'resource-pressure' | 'stability';
}

type RenderDiagnostics = Pick<
  DiagnosticsSnapshot,
  'queuedPageRenders' | 'queuedThumbnailRenders' | 'inflightPageRenders' | 'inflightThumbnailRenders'
>;

const DISPLAY_REFRESH_RATES = [60, 90, 100, 120, 144, 165, 240] as const;
const FRAME_WINDOW_SIZE = 120;
const RECOVERY_EVALUATIONS = 6;

export class AdaptivePerformanceController {
  private frameDurations: number[] = [];
  private previousFrameAt: number | null = null;
  private pendingInputAt: number | null = null;
  private inputLatencies: number[] = [];
  private level: AdaptivePerformanceLevel = 0;
  private recoveryEvaluations = 0;
  private detectedFrameIntervalMs: number | null = null;
  private snapshot: AdaptivePerformanceSnapshot = {
    level: 0,
    detectedRefreshHz: 60,
    targetFrameMs: 1000 / 60,
    p95FrameMs: 0,
    p95InputLatencyMs: 0,
    framePressure: 0,
    renderPressure: 0,
    resourcePressure: 0,
    reason: 'headroom',
  };

  observeFrame(frameAt: number): void {
    if (this.pendingInputAt !== null && frameAt >= this.pendingInputAt) {
      this.inputLatencies.push(frameAt - this.pendingInputAt);
      if (this.inputLatencies.length > FRAME_WINDOW_SIZE) {
        this.inputLatencies.splice(0, this.inputLatencies.length - FRAME_WINDOW_SIZE);
      }
      this.pendingInputAt = null;
    }
    if (this.previousFrameAt !== null) {
      const duration = frameAt - this.previousFrameAt;
      if (Number.isFinite(duration) && duration > 0 && duration <= 100) {
        this.frameDurations.push(duration);
        if (this.frameDurations.length > FRAME_WINDOW_SIZE) {
          this.frameDurations.splice(0, this.frameDurations.length - FRAME_WINDOW_SIZE);
        }
      }
    }
    this.previousFrameAt = frameAt;
  }

  observeInput(inputAt: number): void {
    if (this.pendingInputAt === null || inputAt < this.pendingInputAt) {
      this.pendingInputAt = inputAt;
    }
  }

  resetFrames(): void {
    this.frameDurations = [];
    this.previousFrameAt = null;
    this.pendingInputAt = null;
  }

  evaluate(
    diagnostics: RenderDiagnostics,
    resources: DesktopPerformanceResourcesSnapshot | null,
  ): AdaptivePerformanceSnapshot {
    if (this.frameDurations.length < 12) {
      return this.snapshot;
    }

    const sortedFrames = [...this.frameDurations].sort((a, b) => a - b);
    const observedInterval = percentile(sortedFrames, 0.2);
    const resourceDisplayRefreshHz = resources?.displayRefreshHz;
    const detectedInterval = resourceDisplayRefreshHz && resourceDisplayRefreshHz > 0
      ? 1000 / resourceDisplayRefreshHz
      : resolveDisplayFrameInterval(observedInterval);
    this.detectedFrameIntervalMs = detectedInterval;

    const targetFrameMs = this.detectedFrameIntervalMs;
    const p95FrameMs = percentile(sortedFrames, 0.95);
    const p95InputLatencyMs = this.inputLatencies.length > 0
      ? percentile([...this.inputLatencies].sort((a, b) => a - b), 0.95)
      : 0;
    const targetMissRatio = ratioAbove(this.frameDurations, targetFrameMs * 1.25);
    const sixtyFpsMissRatio = ratioAbove(this.frameDurations, 1000 / 60 * 1.08);
    const framePressure = clamp01(Math.max(
      (p95FrameMs / targetFrameMs - 1) / 1.5,
      targetMissRatio * 2,
      sixtyFpsMissRatio * 3,
      p95InputLatencyMs / 50,
    ));
    const renderPressure = resolveRenderPressure(diagnostics);
    const resourcePressure = resolveResourcePressure(resources);
    const desiredLevel = resolveDesiredPerformanceLevel({
      p95FrameMs,
      targetFrameMs,
      targetMissRatio,
      sixtyFpsMissRatio,
      renderPressure,
      resourcePressure,
      p95InputLatencyMs,
    });

    if (desiredLevel > this.level) {
      this.level = desiredLevel === 3 || p95InputLatencyMs >= 33
        ? desiredLevel
        : nextLevel(this.level);
      this.recoveryEvaluations = 0;
    } else if (desiredLevel < this.level) {
      this.recoveryEvaluations += 1;
      if (this.recoveryEvaluations >= RECOVERY_EVALUATIONS) {
        this.level = previousLevel(this.level);
        this.recoveryEvaluations = 0;
      }
    } else {
      this.recoveryEvaluations = 0;
    }

    const reason = resolvePressureReason({
      level: this.level,
      p95FrameMs,
      renderPressure,
      resourcePressure,
      p95InputLatencyMs,
      targetFrameMs,
    });
    this.snapshot = {
      level: this.level,
      detectedRefreshHz: Math.round(1000 / targetFrameMs),
      targetFrameMs: round(targetFrameMs),
      p95FrameMs: round(p95FrameMs),
      p95InputLatencyMs: round(p95InputLatencyMs),
      framePressure: round(framePressure),
      renderPressure: round(renderPressure),
      resourcePressure: round(resourcePressure),
      reason,
    };
    return this.snapshot;
  }

  current(): AdaptivePerformanceSnapshot {
    return this.snapshot;
  }
}

export function resolveAdaptiveHqDelayMs({
  level,
  isTargetPage,
  viewportInMotion,
  immediateTargetPromotion = false,
}: {
  level: AdaptivePerformanceLevel;
  isTargetPage: boolean;
  viewportInMotion: boolean;
  immediateTargetPromotion?: boolean;
}): number {
  if (immediateTargetPromotion) {
    return 0;
  }

  const targetDelays = viewportInMotion ? [16, 32, 64, 96] : [0, 8, 24, 48];
  const visibleDelays = viewportInMotion ? [48, 72, 112, 160] : [24, 40, 64, 96];
  return (isTargetPage ? targetDelays : visibleDelays)[level]!;
}

export function resolveAdaptiveMotionOverscanPx(level: AdaptivePerformanceLevel): number {
  return [1200, 1200, 900, 600][level]!;
}

export function shouldAllowAdaptivePrefetch(level: AdaptivePerformanceLevel): boolean {
  return level <= 1;
}

export function normalizeAdaptiveInputTimestamp(eventTimestamp: number, currentTime: number): number {
  if (
    !Number.isFinite(eventTimestamp)
    || eventTimestamp <= 0
    || eventTimestamp > currentTime + 1
    || currentTime - eventTimestamp > 1000
  ) {
    return currentTime;
  }
  return eventTimestamp;
}

export function resolveResourcePressure(resources: DesktopPerformanceResourcesSnapshot | null): number {
  if (!resources || resources.totalMemoryKiB <= 0) {
    return 0;
  }
  const reclaimableRatio = resources.reclaimableMemoryKiB / resources.totalMemoryKiB;
  const appMemoryRatio = resources.appWorkingSetKiB / resources.totalMemoryKiB;
  const memoryPressure = reclaimableRatio < 0.04 || appMemoryRatio > 0.35
    ? 1
    : reclaimableRatio < 0.08 || appMemoryRatio > 0.25
      ? 0.7
      : reclaimableRatio < 0.14 || appMemoryRatio > 0.15
        ? 0.35
        : 0;
  const systemCpuPressure = resources.systemCpuUsagePercent === null
    ? 0
    : resources.systemCpuUsagePercent >= 95
      ? 1
      : resources.systemCpuUsagePercent >= 85
        ? 0.7
        : resources.systemCpuUsagePercent >= 70
          ? 0.35
          : 0;
  const appCpuPressure = resources.appCpuPercent >= 300 ? 0.7 : resources.appCpuPercent >= 150 ? 0.35 : 0;
  return Math.max(memoryPressure, systemCpuPressure, appCpuPressure);
}

function resolveDesiredPerformanceLevel({
  p95FrameMs,
  targetFrameMs,
  targetMissRatio,
  sixtyFpsMissRatio,
  renderPressure,
  resourcePressure,
  p95InputLatencyMs,
}: {
  p95FrameMs: number;
  targetFrameMs: number;
  targetMissRatio: number;
  sixtyFpsMissRatio: number;
  renderPressure: number;
  resourcePressure: number;
  p95InputLatencyMs: number;
}): AdaptivePerformanceLevel {
  if (p95FrameMs >= 25 || p95InputLatencyMs >= 50 || sixtyFpsMissRatio >= 0.2 || renderPressure >= 1) {
    return 3;
  }
  if (p95FrameMs > 1000 / 60 * 1.08 || p95InputLatencyMs >= 33 || sixtyFpsMissRatio >= 0.08 || renderPressure >= 0.7 || resourcePressure >= 0.7) {
    return 2;
  }
  if (p95FrameMs > targetFrameMs * 1.2 || p95InputLatencyMs >= Math.max(20, targetFrameMs * 2) || targetMissRatio >= 0.12 || renderPressure >= 0.35 || resourcePressure >= 0.35) {
    return 1;
  }
  return 0;
}

function resolveRenderPressure(diagnostics: RenderDiagnostics): number {
  const queued = diagnostics.queuedPageRenders + diagnostics.queuedThumbnailRenders;
  const inflight = diagnostics.inflightPageRenders + diagnostics.inflightThumbnailRenders;
  if (queued >= 6 || inflight >= 8) return 1;
  if (queued >= 3 || inflight >= 5) return 0.7;
  if (queued >= 1 || inflight >= 3) return 0.35;
  return 0;
}

function resolvePressureReason({
  level,
  p95FrameMs,
  renderPressure,
  resourcePressure,
  p95InputLatencyMs,
  targetFrameMs,
}: {
  level: AdaptivePerformanceLevel;
  p95FrameMs: number;
  renderPressure: number;
  resourcePressure: number;
  p95InputLatencyMs: number;
  targetFrameMs: number;
}): AdaptivePerformanceSnapshot['reason'] {
  if (level === 3) return 'stability';
  if (p95InputLatencyMs >= Math.max(20, targetFrameMs * 2)) return 'input-latency';
  if (resourcePressure >= Math.max(renderPressure, 0.35)) return 'resource-pressure';
  if (renderPressure >= 0.35) return 'render-backlog';
  if (p95FrameMs > targetFrameMs * 1.2) return 'frame-time';
  return 'headroom';
}

function resolveDisplayFrameInterval(observedInterval: number): number {
  let bestInterval = 1000 / DISPLAY_REFRESH_RATES[0];
  let bestDistance = Number.POSITIVE_INFINITY;
  for (const refreshRate of DISPLAY_REFRESH_RATES) {
    const interval = 1000 / refreshRate;
    const distance = Math.abs(interval - observedInterval);
    if (distance < bestDistance) {
      bestDistance = distance;
      bestInterval = interval;
    }
  }
  return bestInterval;
}

function percentile(sorted: readonly number[], ratio: number): number {
  return sorted[Math.min(sorted.length - 1, Math.floor(sorted.length * ratio))] ?? 0;
}

function ratioAbove(values: readonly number[], threshold: number): number {
  return values.filter((value) => value > threshold).length / Math.max(1, values.length);
}

function nextLevel(level: AdaptivePerformanceLevel): AdaptivePerformanceLevel {
  return Math.min(3, level + 1) as AdaptivePerformanceLevel;
}

function previousLevel(level: AdaptivePerformanceLevel): AdaptivePerformanceLevel {
  return Math.max(0, level - 1) as AdaptivePerformanceLevel;
}

function clamp01(value: number): number {
  return Math.max(0, Math.min(1, value));
}

function round(value: number): number {
  return Math.round(value * 1000) / 1000;
}

File: experiments/gpui-migration/gpui-component-compat/src/adaptive_performance.rs
use std::{cmp::Ordering, collections::VecDeque, time::Instant};

const FRAME_WINDOW_SIZE: usize = 120;
const MINIMUM_FRAME_SAMPLES: usize = 12;
const RECOVERY_EVALUATIONS: usize = 6;
const DISPLAY_REFRESH_RATES: [f64; 7] = [60., 90., 100., 120., 144., 165., 240.];

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct ViewerRenderDiagnostics {
    pub queued_page_renders: usize,
    pub queued_thumbnail_renders: usize,
    pub inflight_page_renders: usize,
    pub inflight_thumbnail_renders: usize,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct AdaptiveResourceSnapshot {
    pub total_memory_kib: u64,
    pub reclaimable_memory_kib: u64,
    pub app_working_set_kib: u64,
    pub app_cpu_percent: f64,
    pub system_cpu_usage_percent: Option<f64>,
    pub display_refresh_hz: Option<f64>,
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub enum AdaptivePerformanceReason {
    #[default]
    Headroom,
    FrameTime,
    InputLatency,
    RenderBacklog,
    ResourcePressure,
    Stability,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct AdaptivePerformanceSnapshot {
    pub level: usize,
    pub detected_refresh_hz: u32,
    pub target_frame_ms: f64,
    pub p95_frame_ms: f64,
    pub p95_input_latency_ms: f64,
    pub frame_pressure: f64,
    pub render_pressure: f64,
    pub resource_pressure: f64,
    pub reason: AdaptivePerformanceReason,
}

impl Default for AdaptivePerformanceSnapshot {
    fn default() -> Self {
        Self {
            level: 0,
            detected_refresh_hz: 60,
            target_frame_ms: 1_000. / 60.,
            p95_frame_ms: 0.,
            p95_input_latency_ms: 0.,
            frame_pressure: 0.,
            render_pressure: 0.,
            resource_pressure: 0.,
            reason: AdaptivePerformanceReason::Headroom,
        }
    }
}

/// Application-owned port of Electron's adaptive viewer-performance policy.
///
/// GPUI does not expose the platform event timestamp or a portable resource
/// sampler at the pinned revision. Input latency therefore starts when GPUI
/// receives an event, and callers pass `None` for resources until a reviewed
/// platform capability supplies them.
#[derive(Default)]
pub struct AdaptiveViewerPerformance {
    frame_durations_ms: VecDeque<f64>,
    previous_frame_at: Option<Instant>,
    pending_input_at: Option<Instant>,
    input_latencies_ms: VecDeque<f64>,
    level: usize,
    recovery_evaluations: usize,
    snapshot: AdaptivePerformanceSnapshot,
}

impl AdaptiveViewerPerformance {
    pub fn observe_frame(&mut self, frame_at: Instant) {
        if let Some(input_at) = self.pending_input_at
            && frame_at >= input_at
        {
            push_bounded(
                &mut self.input_latencies_ms,
                frame_at.duration_since(input_at).as_secs_f64() * 1_000.,
            );
            self.pending_input_at = None;
        }
        if let Some(previous) = self.previous_frame_at
            && frame_at >= previous
        {
            let duration_ms = frame_at.duration_since(previous).as_secs_f64() * 1_000.;
            if duration_ms.is_finite() && duration_ms > 0. && duration_ms <= 100. {
                push_bounded(&mut self.frame_durations_ms, duration_ms);
            }
        }
        self.previous_frame_at = Some(frame_at);
    }

    pub fn observe_input(&mut self, input_at: Instant) {
        if self
            .pending_input_at
            .is_none_or(|pending| input_at < pending)
        {
            self.pending_input_at = Some(input_at);
        }
    }

    pub fn reset_frames(&mut self) {
        self.frame_durations_ms.clear();
        self.previous_frame_at = None;
        self.pending_input_at = None;
    }

    pub fn evaluate(
        &mut self,
        diagnostics: ViewerRenderDiagnostics,
    ) -> AdaptivePerformanceSnapshot {
        self.evaluate_with_resources(diagnostics, None)
    }

    pub fn evaluate_with_resources(
        &mut self,
        diagnostics: ViewerRenderDiagnostics,
        resources: Option<AdaptiveResourceSnapshot>,
    ) -> AdaptivePerformanceSnapshot {
        if self.frame_durations_ms.len() < MINIMUM_FRAME_SAMPLES {
            return self.snapshot;
        }

        let mut sorted_frames = self.frame_durations_ms.iter().copied().collect::<Vec<_>>();
        sorted_frames.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Ordering::Equal));
        let observed_interval = percentile(&sorted_frames, 0.2);
        let target_frame_ms = resources
            .and_then(|snapshot| snapshot.display_refresh_hz)
            .filter(|refresh| *refresh > 0.)
            .map(|refresh| 1_000. / refresh)
            .unwrap_or_else(|| resolve_display_frame_interval(observed_interval));
        let p95_frame_ms = percentile(&sorted_frames, 0.95);
        let mut sorted_input = self.input_latencies_ms.iter().copied().collect::<Vec<_>>();
        sorted_input.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Ordering::Equal));
        let p95_input_latency_ms = if sorted_input.is_empty() {
            0.
        } else {
            percentile(&sorted_input, 0.95)
        };
        let target_miss_ratio = ratio_above(&self.frame_durations_ms, target_frame_ms * 1.25);
        let sixty_fps_miss_ratio = ratio_above(&self.frame_durations_ms, 1_000. / 60. * 1.08);
        let frame_pressure = clamp01(
            ((p95_frame_ms / target_frame_ms - 1.) / 1.5)
                .max(target_miss_ratio * 2.)
                .max(sixty_fps_miss_ratio * 3.)
                .max(p95_input_latency_ms / 50.),
        );
        let render_pressure = resolve_render_pressure(diagnostics);
        let resource_pressure = resolve_resource_pressure(resources);
        let desired_level = resolve_desired_level(
            p95_frame_ms,
            target_frame_ms,
            target_miss_ratio,
            sixty_fps_miss_ratio,
            render_pressure,
            resource_pressure,
            p95_input_latency_ms,
        );

        if desired_level > self.level {
            self.level = if desired_level == 3 || p95_input_latency_ms >= 33. {
                desired_level
            } else {
                (self.level + 1).min(3)
            };
            self.recovery_evaluations = 0;
        } else if desired_level < self.level {
            self.recovery_evaluations += 1;
            if self.recovery_evaluations >= RECOVERY_EVALUATIONS {
                self.level = self.level.saturating_sub(1);
                self.recovery_evaluations = 0;
            }
        } else {
            self.recovery_evaluations = 0;
        }

        let reason = if self.level == 3 {
            AdaptivePerformanceReason::Stability
        } else if p95_input_latency_ms >= 20_f64.max(target_frame_ms * 2.) {
            AdaptivePerformanceReason::InputLatency
        } else if resource_pressure >= render_pressure.max(0.35) {
            AdaptivePerformanceReason::ResourcePressure
        } else if render_pressure >= 0.35 {
            AdaptivePerformanceReason::RenderBacklog
        } else if p95_frame_ms > target_frame_ms * 1.2 {
            AdaptivePerformanceReason::FrameTime
        } else {
            AdaptivePerformanceReason::Headroom
        };
        self.snapshot = AdaptivePerformanceSnapshot {
            level: self.level,
            detected_refresh_hz: (1_000. / target_frame_ms).round() as u32,
            target_frame_ms: round(target_frame_ms),
            p95_frame_ms: round(p95_frame_ms),
            p95_input_latency_ms: round(p95_input_latency_ms),
            frame_pressure: round(frame_pressure),
            render_pressure: round(render_pressure),
            resource_pressure: round(resource_pressure),
            reason,
        };
        self.snapshot
    }

    pub const fn current(&self) -> AdaptivePerformanceSnapshot {
        self.snapshot
    }
}

fn push_bounded(values: &mut VecDeque<f64>, value: f64) {
    values.push_back(value);
    while values.len() > FRAME_WINDOW_SIZE {
        values.pop_front();
    }
}

fn resolve_render_pressure(diagnostics: ViewerRenderDiagnostics) -> f64 {
    let queued = diagnostics.queued_page_renders + diagnostics.queued_thumbnail_renders;
    let inflight = diagnostics.inflight_page_renders + diagnostics.inflight_thumbnail_renders;
    if queued >= 6 || inflight >= 8 {
        1.
    } else if queued >= 3 || inflight >= 5 {
        0.7
    } else if queued >= 1 || inflight >= 3 {
        0.35
    } else {
        0.
    }
}

fn resolve_resource_pressure(resources: Option<AdaptiveResourceSnapshot>) -> f64 {
    let Some(resources) = resources.filter(|snapshot| snapshot.total_memory_kib > 0) else {
        return 0.;
    };
    let reclaimable = resources.reclaimable_memory_kib as f64 / resources.total_memory_kib as f64;
    let app_memory = resources.app_working_set_kib as f64 / resources.total_memory_kib as f64;
    let memory: f64 = if reclaimable < 0.04 || app_memory > 0.35 {
        1.
    } else if reclaimable < 0.08 || app_memory > 0.25 {
        0.7
    } else if reclaimable < 0.14 || app_memory > 0.15 {
        0.35
    } else {
        0.
    };
    let system_cpu = resources.system_cpu_usage_percent.map_or(0., |cpu| {
        if cpu >= 95. {
            1.
        } else if cpu >= 85. {
            0.7
        } else if cpu >= 70. {
            0.35
        } else {
            0.
        }
    });
    let app_cpu = if resources.app_cpu_percent >= 300. {
        0.7
    } else if resources.app_cpu_percent >= 150. {
        0.35
    } else {
        0.
    };
    memory.max(system_cpu).max(app_cpu)
}

#[allow(clippy::too_many_arguments)]
fn resolve_desired_level(
    p95_frame_ms: f64,
    target_frame_ms: f64,
    target_miss_ratio: f64,
    sixty_fps_miss_ratio: f64,
    render_pressure: f64,
    resource_pressure: f64,
    p95_input_latency_ms: f64,
) -> usize {
    if p95_frame_ms >= 25.
        || p95_input_latency_ms >= 50.
        || sixty_fps_miss_ratio >= 0.2
        || render_pressure >= 1.
    {
        3
    } else if p95_frame_ms > 1_000. / 60. * 1.08
        || p95_input_latency_ms >= 33.
        || sixty_fps_miss_ratio >= 0.08
        || render_pressure >= 0.7
        || resource_pressure >= 0.7
    {
        2
    } else if p95_frame_ms > target_frame_ms * 1.2
        || p95_input_latency_ms >= 20_f64.max(target_frame_ms * 2.)
        || target_miss_ratio >= 0.12
        || render_pressure >= 0.35
        || resource_pressure >= 0.35
    {
        1
    } else {
        0
    }
}

fn resolve_display_frame_interval(observed_interval: f64) -> f64 {
    DISPLAY_REFRESH_RATES
        .into_iter()
        .map(|refresh| 1_000. / refresh)
        .min_by(|a, b| {
            (a - observed_interval)
                .abs()
                .partial_cmp(&(b - observed_interval).abs())
                .unwrap_or(Ordering::Equal)
        })
        .unwrap_or(1_000. / 60.)
}

fn percentile(sorted: &[f64], ratio: f64) -> f64 {
    sorted[((sorted.len() as f64 * ratio).floor() as usize).min(sorted.len() - 1)]
}

fn ratio_above(values: &VecDeque<f64>, threshold: f64) -> f64 {
    values.iter().filter(|value| **value > threshold).count() as f64 / values.len().max(1) as f64
}

fn clamp01(value: f64) -> f64 {
    value.clamp(0., 1.)
}

fn round(value: f64) -> f64 {
    (value * 1_000.).round() / 1_000.
}
